File: backend/src/routes/projects.ts
import { Router, Request, Response } from 'express';
import { PrismaClient } from '@prisma/client';
import { getPaginationParams, buildPaginatedResponse } from '../middleware/pagination';

const prisma = new PrismaClient();
const router = Router();

function mapProject(p: any) {
  return {
    id:                p.id,
    projectId:         p.project_id,
    name:              p.project_name,
    projectName:       p.project_name,
    type:              p.type,
    startDate:         p.start_date,
    goLiveDate:        p.go_live_date,
    ppsExitDate:       p.pps_exit_date,
    endDate:           p.end_date,
    pmName:            p.pm,
    pm:                p.pm,
    pmItcode:          p.pm_itcode,
    dtLeadName:        p.dt_lead,
    dtLead:            p.dt_lead,
    dtLeadItcode:      p.dt_lead_itcode,
    itLeadName:        p.it_lead,
    itLead:            p.it_lead,
    itLeadItcode:      p.it_lead_itcode,
    duration:          p.duration,
    objectives:        p.objectives,
    investmentCost:    p.investment_cost,
    yearlyMaCost:      p.yearly_ma_cost,
    currency:          p.currency,
    expectedManDays:   p.expected_man_days,
    status:            p.status,
    comment:           p.comment,
    eaReviewType:      p.ea_review_type,
    domainEaReviewer:  p.domain_ea_reviewer,
    favourite:         p.favourite,
    overallStatus:     p.overall_status,
    approvedTime:      p.approved_time,
    source:            p.source,
    eaApprovalDt:      p.ea_approval_dt,
    aiRelated:         p.ai_related,
    createdBy:         p.create_by,
    createdAt:         p.create_at,
    updatedAt:         p.update_at,
  };
}

// GET /api/projects — List with pagination, filtering, sorting
router.get('/', async (req: Request, res: Response) => {
  try {
    const { page, pageSize, sortField, sortOrder, skip } = getPaginationParams(req);
    const { projectId, name, itCode, requestStatus, aiRelated } = req.query;

    const where: any = {};
    if (projectId)     where.project_id   = { contains: projectId as string, mode: 'insensitive' };
    if (name)          where.project_name = { contains: name as string, mode: 'insensitive' };
    if (itCode)        where.it_lead      = { contains: itCode as string, mode: 'insensitive' };
    if (requestStatus) where.status       = requestStatus as string;
    if (aiRelated)     where.ai_related   = aiRelated === 'true' ? 'Yes' : 'No';

    const fieldMap: Record<string, string> = {
      projectId:  'project_id',
      name:       'project_name',
      pmName:     'pm',
      dtLeadName: 'dt_lead',
      itLeadName: 'it_lead',
      createdAt:  'create_at',
      aiRelated:  'ai_related',
    };
    const dbSortField = sortField ? (fieldMap[sortField] || sortField) : 'project_id';
    const orderBy: any = { [dbSortField]: sortOrder || 'desc' };

    const [data, total] = await Promise.all([
      prisma.project.findMany({ where, orderBy, skip, take: pageSize }),
      prisma.project.count({ where }),
    ]);

    res.json(buildPaginatedResponse(data.map(mapProject), total, page, pageSize));
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to fetch projects' });
  }
});

// GET /api/projects/:id — Get single project by project_id
router.get('/:id', async (req: Request, res: Response) => {
  try {
    const project = await prisma.project.findFirst({ where: { project_id: req.params.id as string } });
    if (!project) return res.status(404).json({ error: 'Project not found' });
    res.json(mapProject(project));
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to fetch project' });
  }
});

// POST /api/projects — Create new project
router.post('/', async (req: Request, res: Response) => {
  try {
    const {
      projectName, type, startDate, goLiveDate, ppsExitDate, endDate,
      pm, pmItcode, dtLead, dtLeadItcode, itLead, itLeadItcode,
      duration, objectives, investmentCost, yearlyMaCost, currency,
      expectedManDays, comment, eaReviewType, domainEaReviewer,
      aiRelated, source, createdBy,
    } = req.body;

    const project = await prisma.project.create({
      data: {
        project_name:       projectName ?? null,
        type:               type ?? null,
        start_date:         startDate ?? null,
        go_live_date:       goLiveDate ?? null,
        pps_exit_date:      ppsExitDate ?? null,
        end_date:           endDate ?? null,
        pm:                 pm ?? null,
        pm_itcode:          pmItcode ?? null,
        dt_lead:            dtLead ?? null,
        dt_lead_itcode:     dtLeadItcode ?? null,
        it_lead:            itLead ?? null,
        it_lead_itcode:     itLeadItcode ?? null,
        duration:           duration ?? null,
        objectives:         objectives ?? null,
        investment_cost:    investmentCost ?? null,
        yearly_ma_cost:     yearlyMaCost ?? null,
        currency:           currency ?? null,
        expected_man_days:  expectedManDays ?? null,
        status:             'Active',
        comment:            comment ?? null,
        ea_review_type:     eaReviewType ?? null,
        domain_ea_reviewer: domainEaReviewer ?? null,
        ai_related:         aiRelated ?? null,
        source:             source ?? null,
        create_by:          createdBy ?? null,
      },
    });

    res.status(201).json(mapProject(project));
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to create project' });
  }
});

// PUT /api/projects/:id — Update project
router.put('/:id', async (req: Request, res: Response) => {
  try {
    const existing = await prisma.project.findFirst({ where: { project_id: req.params.id as string } });
    if (!existing) return res.status(404).json({ error: 'Project not found' });

    const {
      projectName, type, startDate, goLiveDate, ppsExitDate, endDate,
      pm, pmItcode, dtLead, dtLeadItcode, itLead, itLeadItcode,
      duration, objectives, investmentCost, yearlyMaCost, currency,
      expectedManDays, status, comment, eaReviewType, domainEaReviewer,
      favourite, overallStatus, aiRelated, source, updatedBy,
    } = req.body;

    const updateData: any = { update_at: new Date() };
    if (projectName !== undefined)       updateData.project_name       = projectName;
    if (type !== undefined)              updateData.type               = type;
    if (startDate !== undefined)         updateData.start_date         = startDate;
    if (goLiveDate !== undefined)        updateData.go_live_date       = goLiveDate;
    if (ppsExitDate !== undefined)       updateData.pps_exit_date      = ppsExitDate;
    if (endDate !== undefined)           updateData.end_date           = endDate;
    if (pm !== undefined)                updateData.pm                 = pm;
    if (pmItcode !== undefined)          updateData.pm_itcode          = pmItcode;
    if (dtLead !== undefined)            updateData.dt_lead            = dtLead;
    if (dtLeadItcode !== undefined)      updateData.dt_lead_itcode     = dtLeadItcode;
    if (itLead !== undefined)            updateData.it_lead            = itLead;
    if (itLeadItcode !== undefined)      updateData.it_lead_itcode     = itLeadItcode;
    if (duration !== undefined)          updateData.duration           = duration;
    if (objectives !== undefined)        updateData.objectives         = objectives;
    if (investmentCost !== undefined)    updateData.investment_cost    = investmentCost;
    if (yearlyMaCost !== undefined)      updateData.yearly_ma_cost     = yearlyMaCost;
    if (currency !== undefined)          updateData.currency           = currency;
    if (expectedManDays !== undefined)   updateData.expected_man_days  = expectedManDays;
    if (status !== undefined)            updateData.status             = status;
    if (comment !== undefined)           updateData.comment            = comment;
    if (eaReviewType !== undefined)      updateData.ea_review_type     = eaReviewType;
    if (domainEaReviewer !== undefined)  updateData.domain_ea_reviewer = domainEaReviewer;
    if (favourite !== undefined)         updateData.favourite          = favourite;
    if (overallStatus !== undefined)     updateData.overall_status     = overallStatus;
    if (aiRelated !== undefined)         updateData.ai_related         = aiRelated;
    if (source !== undefined)            updateData.source             = source;
    if (updatedBy !== undefined)         updateData.update_by          = updatedBy;

    const updated = await prisma.project.update({
      where: { id: existing.id },
      data: updateData,
    });

    res.json(mapProject(updated));
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to update project' });
  }
});

export { router as projectRoutes };

File: backend/app/routers/projects.py
"""Projects router — CRUD with pagination, filtering, sorting."""
from fastapi import APIRouter, Depends, Query, HTTPException
from fastapi.responses import JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import text

from app.database import get_db
from app.utils.pagination import PaginationParams, paginated_response
from app.utils.filters import multi_value_condition

router = APIRouter()

# ── Mapping helpers ──────────────────────────────────────────────

SORT_FIELD_MAP: dict[str, str] = {
    "projectId": "project_id",
    "name": "project_name",
    "pmName": "pm",
    "dtLeadName": "dt_lead",
    "itLeadName": "it_lead",
    "createdAt": "create_at",
    "aiRelated": "ai_related",
}

ALLOWED_SORT_FIELDS = set(SORT_FIELD_MAP.values()) | {
    "project_id", "project_name", "pm", "dt_lead", "it_lead",
    "create_at", "ai_related",
}


def _map_project(row) -> dict:
    """Convert a DB row (RowMapping) to the camelCase API shape."""
    r = row._mapping if hasattr(row, "_mapping") else row
    return {
        "id": r["id"],
        "projectId": r["project_id"],
        "name": r["project_name"],
        "projectName": r["project_name"],
        "type": r["type"],
        "startDate": r["start_date"],
        "goLiveDate": r["go_live_date"],
        "ppsExitDate": r["pps_exit_date"],
        "endDate": r["end_date"],
        "pmName": r["pm"],
        "pm": r["pm"],
        "pmItcode": r["pm_itcode"],
        "dtLeadName": r["dt_lead"],
        "dtLead": r["dt_lead"],
        "dtLeadItcode": r["dt_lead_itcode"],
        "itLeadName": r["it_lead"],
        "itLead": r["it_lead"],
        "itLeadItcode": r["it_lead_itcode"],
        "duration": r["duration"],
        "objectives": r["objectives"],
        "investmentCost": r["investment_cost"],
        "yearlyMaCost": r["yearly_ma_cost"],
        "currency": r["currency"],
        "expectedManDays": r["expected_man_days"],
        "status": r["status"],
        "comment": r["comment"],
        "eaReviewType": r["ea_review_type"],
        "domainEaReviewer": r["domain_ea_reviewer"],
        "favourite": r["favourite"],
        "overallStatus": r["overall_status"],
        "approvedTime": r["approved_time"],
        "source": r["source"],
        "eaApprovalDt": r["ea_approval_dt"],
        "aiRelated": r["ai_related"],
        "createdBy": r["create_by"],
        "createdAt": r["create_at"],
        "updatedAt": r["update_at"],
    }


# ── GET /api/projects ────────────────────────────────────────────

@router.get("")
async def list_projects(
    pagination: PaginationParams = Depends(),
    projectId: str | None = Query(None),
    name: str | None = Query(None),
    itCode: str | None = Query(None),
    requestStatus: str | None = Query(None),
    aiRelated: str | None = Query(None),
    db: AsyncSession = Depends(get_db),
):
    try:
        conditions: list[str] = []
        params: dict = {}

        if projectId:
            conditions.append("project_id ILIKE :projectId")
            params["projectId"] = f"%{projectId}%"
        if name:
            conditions.append("project_name ILIKE :name")
            params["name"] = f"%{name}%"
        if itCode:
            conditions.append("it_lead ILIKE :itCode")
            params["itCode"] = f"%{itCode}%"
        if requestStatus:
            conditions.append(multi_value_condition("status", "requestStatus", requestStatus, params))
        if aiRelated:
            ai_vals = [v.strip() for v in aiRelated.split(",") if v.strip()]
            mapped = ["Yes" if v.lower() in ("true", "yes") else "No" for v in ai_vals]
            conditions.append(multi_value_condition("ai_related", "aiRelated", ",".join(mapped), params))

        where_clause = " AND ".join(conditions) if conditions else "1=1"

        # Sort
        db_sort_field = SORT_FIELD_MAP.get(pagination.sort_field, pagination.sort_field) if pagination.sort_field else "project_id"
        if db_sort_field not in ALLOWED_SORT_FIELDS:
            db_sort_field = "project_id"
        sort_order = "ASC" if pagination.sort_order and pagination.sort_order.lower() == "asc" else "DESC"

        data_query = text(
            f"SELECT * FROM eam.project WHERE {where_clause} "
            f"ORDER BY {db_sort_field} {sort_order} "
            f"LIMIT :limit OFFSET :offset"
        )
        count_query = text(
            f"SELECT COUNT(*) FROM eam.project WHERE {where_clause}"
        )

        params["limit"] = pagination.page_size
        params["offset"] = pagination.offset

        result = await db.execute(data_query, params)
        rows = result.mappings().all()

        count_result = await db.execute(count_query, {k: v for k, v in params.items() if k not in ("limit", "offset")})
        total = count_result.scalar_one()

        return paginated_response(
            [_map_project(r) for r in rows],
            total,
            pagination.page,
            pagination.page_size,
        )
    except Exception as e:
        print(f"Error fetching projects: {e}")
        raise HTTPException(status_code=500, detail="Failed to fetch projects")


# ── GET /api/projects/{project_id} ───────────────────────────────

@router.get("/{project_id}")
async def get_project(
    project_id: str,
    db: AsyncSession = Depends(get_db),
):
    try:
        result = await db.execute(
            text("SELECT * FROM eam.project WHERE project_id = :pid"),
            {"pid": project_id},
        )
        row = result.mappings().first()
        if not row:
            raise HTTPException(status_code=404, detail="Project not found")
        return _map_project(row)
    except HTTPException:
        raise
    except Exception as e:
        print(f"Error fetching project: {e}")
        raise HTTPException(status_code=500, detail="Failed to fetch project")


# ── POST /api/projects ───────────────────────────────────────────

@router.post("", status_code=201)
async def create_project(
    body: dict,
    db: AsyncSession = Depends(get_db),
):
    try:
        result = await db.execute(
            text("""
                INSERT INTO eam.project (
                    project_name, type, start_date, go_live_date, pps_exit_date, end_date,
                    pm, pm_itcode, dt_lead, dt_lead_itcode, it_lead, it_lead_itcode,
                    duration, objectives, investment_cost, yearly_ma_cost, currency,
                    expected_man_days, status, comment, ea_review_type, domain_ea_reviewer,
                    ai_related, source, create_by
                ) VALUES (
                    :project_name, :type, :start_date, :go_live_date, :pps_exit_date, :end_date,
                    :pm, :pm_itcode, :dt_lead, :dt_lead_itcode, :it_lead, :it_lead_itcode,
                    :duration, :objectives, :investment_cost, :yearly_ma_cost, :currency,
                    :expected_man_days, 'Active', :comment, :ea_review_type, :domain_ea_reviewer,
                    :ai_related, :source, :create_by
                ) RETURNING *
            """),
            {
                "project_name": body.get("projectName"),
                "type": body.get("type"),
                "start_date": body.get("startDate"),
                "go_live_date": body.get("goLiveDate"),
                "pps_exit_date": body.get("ppsExitDate"),
                "end_date": body.get("endDate"),
                "pm": body.get("pm"),
                "pm_itcode": body.get("pmItcode"),
                "dt_lead": body.get("dtLead"),
                "dt_lead_itcode": body.get("dtLeadItcode"),
                "it_lead": body.get("itLead"),
                "it_lead_itcode": body.get("itLeadItcode"),
                "duration": body.get("duration"),
                "objectives": body.get("objectives"),
                "investment_cost": body.get("investmentCost"),
                "yearly_ma_cost": body.get("yearlyMaCost"),
                "currency": body.get("currency"),
                "expected_man_days": body.get("expectedManDays"),
                "comment": body.get("comment"),
                "ea_review_type": body.get("eaReviewType"),
                "domain_ea_reviewer": body.get("domainEaReviewer"),
                "ai_related": body.get("aiRelated"),
                "source": body.get("source"),
                "create_by": body.get("createdBy"),
            },
        )
        row = result.mappings().first()
        await db.commit()
        return _map_project(row)
    except Exception as e:
        await db.rollback()
        print(f"Error creating project: {e}")
        raise HTTPException(status_code=500, detail="Failed to create project")


# ── PUT /api/projects/{project_id} ───────────────────────────────

@router.put("/{project_id}")
async def update_project(
    project_id: str,
    body: dict,
    db: AsyncSession = Depends(get_db),
):
    try:
        # Check existence
        existing = await db.execute(
            text("SELECT id FROM eam.project WHERE project_id = :pid"),
            {"pid": project_id},
        )
        row = existing.mappings().first()
        if not row:
            raise HTTPException(status_code=404, detail="Project not found")

        db_id = row["id"]

        # Build dynamic SET clause
        field_mapping = {
            "projectName": "project_name",
            "type": "type",
            "startDate": "start_date",
            "goLiveDate": "go_live_date",
            "ppsExitDate": "pps_exit_date",
            "endDate": "end_date",
            "pm": "pm",
            "pmItcode": "pm_itcode",
            "dtLead": "dt_lead",
            "dtLeadItcode": "dt_lead_itcode",
            "itLead": "it_lead",
            "itLeadItcode": "it_lead_itcode",
            "duration": "duration",
            "objectives": "objectives",
            "investmentCost": "investment_cost",
            "yearlyMaCost": "yearly_ma_cost",
            "currency": "currency",
            "expectedManDays": "expected_man_days",
            "status": "status",
            "comment": "comment",
            "eaReviewType": "ea_review_type",
            "domainEaReviewer": "domain_ea_reviewer",
            "favourite": "favourite",
            "overallStatus": "overall_status",
            "aiRelated": "ai_related",
            "source": "source",
            "updatedBy": "update_by",
        }

        set_parts = ["update_at = NOW()"]
        params: dict = {"db_id": db_id}

        for api_field, db_col in field_mapping.items():
            if api_field in body:
                param_name = f"u_{db_col}"
                set_parts.append(f"{db_col} = :{param_name}")
                params[param_name] = body[api_field]

        set_clause = ", ".join(set_parts)

        result = await db.execute(
            text(f"UPDATE eam.project SET {set_clause} WHERE id = :db_id RETURNING *"),
            params,
        )
        updated = result.mappings().first()
        await db.commit()
        return _map_project(updated)
    except HTTPException:
        raise
    except Exception as e:
        await db.rollback()
        print(f"Error updating project: {e}")
        raise HTTPException(status_code=500, detail="Failed to update project")
